File: docx-core/src/documents/elements/style.rs
use serde::Serialize;

use crate::documents::BuildXML;
use crate::types::*;
use crate::xml_builder::*;
use crate::StyleType;

use super::*;

#[derive(Debug, Clone, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Style {
    pub style_id: String,
    pub name: Name,
    pub style_type: StyleType,
    pub run_property: RunProperty,
    pub paragraph_property: ParagraphProperty,
    pub table_property: TableProperty,
    pub table_cell_property: TableCellProperty,
    pub based_on: Option<BasedOn>,
    pub next: Option<Next>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub link: Option<Link>,
}

impl Default for Style {
    fn default() -> Self {
        let name = Name::new("");
        let rpr = RunProperty::new();
        let ppr = ParagraphProperty::new();
        Style {
            style_id: "".to_owned(),
            style_type: StyleType::Paragraph,
            name,
            run_property: rpr,
            paragraph_property: ppr,
            table_property: TableProperty::new(),
            table_cell_property: TableCellProperty::new(),
            based_on: None,
            next: None,
            link: None,
        }
    }
}

impl Style {
    pub fn new(style_id: impl Into<String>, style_type: StyleType) -> Self {
        let default = Default::default();
        Style {
            style_id: style_id.into(),
            style_type,
            ..default
        }
    }

    pub fn name(mut self, name: impl Into<String>) -> Self {
        self.name = Name::new(name);
        self
    }

    pub fn based_on(mut self, base: impl Into<String>) -> Self {
        self.based_on = Some(BasedOn::new(base));
        self
    }

    pub fn next(mut self, next: impl Into<String>) -> Self {
        self.next = Some(Next::new(next));
        self
    }

    pub fn link(mut self, link: impl Into<String>) -> Self {
        self.link = Some(Link::new(link));
        self
    }

    pub fn size(mut self, size: usize) -> Self {
        self.run_property = self.run_property.size(size);
        self
    }

    pub fn color(mut self, color: impl Into<String>) -> Self {
        self.run_property = self.run_property.color(color);
        self
    }

    pub fn highlight(mut self, color: impl Into<String>) -> Self {
        self.run_property = self.run_property.highlight(color);
        self
    }

    pub fn bold(mut self) -> Self {
        self.run_property = self.run_property.bold();
        self
    }

    pub fn italic(mut self) -> Self {
        self.run_property = self.run_property.italic();
        self
    }

    pub fn underline(mut self, line_type: impl Into<String>) -> Self {
        self.run_property = self.run_property.underline(line_type);
        self
    }

    pub fn vanish(mut self) -> Self {
        self.run_property = self.run_property.vanish();
        self
    }

    pub fn text_border(mut self, b: TextBorder) -> Self {
        self.run_property = self.run_property.text_border(b);
        self
    }

    pub fn fonts(mut self, f: RunFonts) -> Self {
        self.run_property = self.run_property.fonts(f);
        self
    }

    pub fn align(mut self, alignment_type: AlignmentType) -> Self {
        self.paragraph_property = self.paragraph_property.align(alignment_type);
        self
    }

    pub fn indent(
        mut self,
        left: Option<i32>,
        special_indent: Option<SpecialIndentType>,
        end: Option<i32>,
        start_chars: Option<i32>,
    ) -> Self {
        self.paragraph_property =
            self.paragraph_property
                .indent(left, special_indent, end, start_chars);
        self
    }

    pub fn hanging_chars(mut self, chars: i32) -> Self {
        self.paragraph_property = self.paragraph_property.hanging_chars(chars);
        self
    }

    pub fn first_line_chars(mut self, chars: i32) -> Self {
        self.paragraph_property = self.paragraph_property.first_line_chars(chars);
        self
    }

    pub fn outline_lvl(mut self, l: usize) -> Self {
        self.paragraph_property = self.paragraph_property.outline_lvl(l);
        self
    }

    pub fn table_property(mut self, p: TableProperty) -> Self {
        self.table_property = p;
        self
    }

    pub fn table_indent(mut self, v: i32) -> Self {
        self.table_property = self.table_property.indent(v);
        self
    }

    pub fn table_align(mut self, v: TableAlignmentType) -> Self {
        self.table_property = self.table_property.align(v);
        self
    }

    pub fn style(mut self, s: impl Into<String>) -> Self {
        self.table_property = self.table_property.style(s);
        self
    }

    pub fn layout(mut self, t: TableLayoutType) -> Self {
        self.table_property = self.table_property.layout(t);
        self
    }

    pub fn width(mut self, w: usize, t: WidthType) -> Self {
        self.table_property = self.table_property.width(w, t);
        self
    }

    pub fn margins(mut self, margins: TableCellMargins) -> Self {
        self.table_property = self.table_property.set_margins(margins);
        self
    }

    pub fn set_borders(mut self, borders: TableBorders) -> Self {
        self.table_property = self.table_property.set_borders(borders);
        self
    }

    pub fn set_border(mut self, border: TableBorder) -> Self {
        self.table_property = self.table_property.set_border(border);
        self
    }

    pub fn clear_border(mut self, position: TableBorderPosition) -> Self {
        self.table_property = self.table_property.clear_border(position);
        self
    }

    pub fn clear_all_border(mut self) -> Self {
        self.table_property = self.table_property.clear_all_border();
        self
    }

    pub fn table_cell_property(mut self, p: TableCellProperty) -> Self {
        self.table_cell_property = p;
        self
    }
}

impl BuildXML for Style {
    fn build(&self) -> Vec<u8> {
        let b = XMLBuilder::new();
        // Set "Normal" as default if you need change these values please fix it
        let mut b = b
            .open_style(self.style_type, &self.style_id)
            .add_child(&self.name)
            .add_child(&self.run_property)
            .add_child(&self.paragraph_property);

        if self.style_type == StyleType::Table {
            b = b
                .add_child(&self.table_cell_property)
                .add_child(&self.table_property);
        }

        if let Some(ref next) = self.next {
            b = b.add_child(next)
        }

        if let Some(ref link) = self.link {
            b = b.add_child(link)
        }

        b.add_child(&QFormat::new())
            .add_optional_child(&self.based_on)
            .close()
            .build()
    }
}

#[cfg(test)]
mod tests {

    use super::*;
    #[cfg(test)]
    use pretty_assertions::assert_eq;
    use std::str;

    #[test]
    fn test_build() {
        let c = Style::new("Heading", StyleType::Paragraph).name("Heading1");
        let b = c.build();
        assert_eq!(
            str::from_utf8(&b).unwrap(),
            r#"<w:style w:type="paragraph" w:styleId="Heading"><w:name w:val="Heading1" /><w:rPr /><w:pPr><w:rPr /></w:pPr><w:qFormat /></w:style>"#
        );
    }
}

File: docx-wasm/js/style.ts
import * as wasm from "./pkg";

import {
  createDefaultTableCellMargins,
  setTableProperty,
  TableAlignmentType,
  TableLayoutType,
  TableProperty,
} from "./table";
import {
  RunProperty,
  createDefaultRunProperty,
  VertAlignType,
  RunFonts,
  setRunProperty,
} from "./run-property";
import {
  AlignmentType,
  createDefaultParagraphProperty,
  createParagraphProperty,
  LineSpacing,
  ParagraphProperty,
  SpecialIndentKind,
} from "./paragraph-property";
import { BorderType } from "./border";
import { WidthType } from ".";

export type StyleType = "paragraph" | "character" | "numbering" | "table";

export class Style {
  _styleId: string;
  _name: string;
  _styleType: StyleType;
  _runProperty: RunProperty;
  _paragraphProperty: ParagraphProperty;
  _tableProperty: TableProperty;
  _basedOn: string | null;
  _link: string | null;

  constructor(id: string, type: StyleType) {
    this._styleId = id;
    this._styleType = type;
    this._name = "";
    this._runProperty;
    this._tableProperty = { cellMargins: createDefaultTableCellMargins() };
    this._runProperty = createDefaultRunProperty();
    this._paragraphProperty = createDefaultParagraphProperty();
    this._basedOn = null;
    this._link = null;
  }

  name = (n: string) => {
    this._name = n;
    return this;
  };

  basedOn = (n: string) => {
    this._basedOn = n;
    return this;
  };

  link = (n: string) => {
    this._link = n;
    return this;
  };

  // run property
  style(style: string) {
    this._runProperty.style(style);
    return this;
  }

  size(size: number) {
    this._runProperty.size(size);
    return this;
  }

  color(color: string) {
    this._runProperty.color(color);
    return this;
  }

  highlight(color: string) {
    this._runProperty.highlight(color);
    return this;
  }

  vertAlign(vertAlign: VertAlignType) {
    this._runProperty.vertAlign(vertAlign);
    return this;
  }

  bold() {
    this._runProperty.bold();
    return this;
  }

  strike() {
    this._runProperty.strike();
    return this;
  }

  dstrike() {
    this._runProperty.dstrike();
    return this;
  }

  italic() {
    this._runProperty.italic();
    return this;
  }

  underline(type: string) {
    this._runProperty.underline(type);
    return this;
  }

  vanish() {
    this._runProperty.vanish();
    return this;
  }

  fonts(fonts: RunFonts) {
    this._runProperty.fonts(fonts);
    return this;
  }

  characterSpacing(characterSpacing: number) {
    this._runProperty.spacing(characterSpacing);
    return this;
  }

  delete(author: string, date: string) {
    this._runProperty.delete(author, date);
    return this;
  }

  insert(author: string, date: string) {
    this._runProperty.insert(author, date);
    return this;
  }

  textBorder(type: BorderType, size: number, space: number, color: string) {
    this._runProperty.textBorder(type, size, space, color);
    return this;
  }

  paragraphProperty = (p: ParagraphProperty) => {
    this._paragraphProperty = p;
    return this;
  };

  runProperty = (p: RunProperty) => {
    this._runProperty = p;
    return this;
  };

  // paragraph property
  align(type: AlignmentType) {
    this._paragraphProperty._align = type;
    return this;
  }

  indent(
    left: number,
    specialIndentKind?: SpecialIndentKind,
    specialIndentSize?: number
  ) {
    this._paragraphProperty.indent = {
      left,
      specialIndentKind,
      specialIndentSize,
    };
    return this;
  }

  numbering(id: number, level: number) {
    this._paragraphProperty.numbering = { id, level };
    return this;
  }

  lineSpacing(spacing: LineSpacing) {
    this._paragraphProperty.lineSpacing = spacing;
    return this;
  }

  keepNext(v: boolean) {
    this._paragraphProperty.keepNext = v;
    return this;
  }

  keepLines(v: boolean) {
    this._paragraphProperty.keepLines = v;
    return this;
  }

  snapToGrid(v: boolean) {
    this._paragraphProperty._snapToGrid = v;
    return this;
  }

  pageBreakBefore(v: boolean) {
    this._paragraphProperty.pageBreakBefore = v;
    return this;
  }

  widowControl(v: boolean) {
    this._paragraphProperty.widowControl = v;
    return this;
  }

  outlineLevel(v: number) {
    this._paragraphProperty.outlineLvl = v;
    return this;
  }

  // TODO:
  // tableProperty = (n: TableProperty) => {
  //   this._tableProperty = n;
  //   return this;
  // };

  // tableProperty
  tableIndent(v: number) {
    this._tableProperty.indent = v;
    return this;
  }

  tableAlign(v: TableAlignmentType) {
    this._tableProperty.align = v;
    return this;
  }

  layout(l: TableLayoutType) {
    this._tableProperty.layout = l;
    return this;
  }

  width(w: number) {
    this._tableProperty.width = w;
    return this;
  }

  cellMargins(top: number, right: number, bottom: number, left: number) {
    this._tableProperty.cellMargins = {
      top: { val: top, type: "dxa" },
      left: { val: left, type: "dxa" },
      bottom: { val: bottom, type: "dxa" },
      right: { val: right, type: "dxa" },
    };
    return this;
  }

  cellMarginTop(v: number, t: WidthType) {
    this._tableProperty.cellMargins.top = { val: v, type: t };
    return this;
  }

  cellMarginLeft(v: number, t: WidthType) {
    this._tableProperty.cellMargins.left = { val: v, type: t };
    return this;
  }

  cellMarginRight(v: number, t: WidthType) {
    this._tableProperty.cellMargins.right = { val: v, type: t };
    return this;
  }

  cellMarginBottom(v: number, t: WidthType) {
    this._tableProperty.cellMargins.bottom = { val: v, type: t };
    return this;
  }

  buildStyleType = () => {
    switch (this._styleType) {
      case "character":
        return wasm.StyleType.Character;
      case "numbering":
        return wasm.StyleType.Numbering;
      case "paragraph":
        return wasm.StyleType.Paragraph;
      case "table":
        return wasm.StyleType.Table;
    }
    return wasm.StyleType.Paragraph;
  };

  buildWasmObject = () => {
    const styleType = this.buildStyleType();

    let s = wasm.createStyle(this._styleId, styleType);

    if (this._name) {
      s = s.name(this._name);
    }

    if (this._basedOn) {
      s = s.based_on(this._basedOn);
    }

    if (this._link) {
      s = s.link(this._link);
    }

    if (this._runProperty) {
      s = setRunProperty(s, this._runProperty);
    }

    if (this._paragraphProperty) {
      s = s.paragraph_property(
        createParagraphProperty(this._paragraphProperty)
      );
    }

    if (this._tableProperty) {
      s = setTableProperty(s, this._tableProperty);
    }

    return s;
  };
}
